src/main/java/dev/meyi/noscamspam/checks/UserCheck.java

```diff
@@ -13,40 +18,51 @@
-      return false;
-    } else if (NoScamSpam.blacklist.contains(username)) {
-      return true;
-    }
+    private static final CloseableHttpClient httpClient = HttpClientBuilder.create().build();
 
-    try (CloseableHttpResponse response = httpClient.execute(new HttpGet(
-        "https://api.hypixel.net/player?key=" + NoScamSpam.config.apiKey + "&name=" + username));
-        InputStream stream = response.getEntity().getContent()) {
-      StatsResponse statsResponse = NoScamSpam.gson
-          .fromJson(IOUtils
-              .toString(stream,
-                  StandardCharsets.UTF_8), StatsResponse.class);
-
-      if (statsResponse.success) {
-        if (statsResponse.player.networkExp > NoScamSpam.config.hypixelExperience &&
-            statsResponse.player.achievements.getSkillAverage() > NoScamSpam.config.skillAverage &&
-            statsResponse.player.achievements.skyblock_dungeoneer
-                > NoScamSpam.config.catacombsLevel) {
-          return false;
-        } else {
-          if (!NoScamSpam.blacklist.contains(username)) {
-            NoScamSpam.blacklist.add(username);
-          }
-          return true;
+    public static void shouldBlockUser(String username, ClientChatReceivedEvent event) {
+        if (NoScamSpam.config.whitelist.contains(username) || NoScamSpam.cachedAllowed.contains(username))
+            return;
+        if (NoScamSpam.blacklist.contains(username)) {
+            event.setCanceled(true);
+            return;
         }
-      }
-    } catch (IOException e) {
-      e.printStackTrace();
-    }
 
-    return !NoScamSpam.config.showOnError;
-  }
+        event.setCanceled(true);
+        Thread fetchThread = new Thread(() -> {
+            try (CloseableHttpResponse response = httpClient.execute(new HttpGet(
+                    "https://api.hypixel.net/player?key=" + NoScamSpam.config.apiKey + "&name=" + username));
+                 InputStream stream = response.getEntity().getContent()) {
+                StatsResponse statsResponse = NoScamSpam.gson
+                        .fromJson(IOUtils
+                                .toString(stream,
+                                        StandardCharsets.UTF_8), StatsResponse.class);
+
+                if (statsResponse.success) {
+                    if (statsResponse.player.networkExp > NoScamSpam.config.hypixelExperience &&
+                            statsResponse.player.achievements.getSkillAverage() > NoScamSpam.config.skillAverage &&
+                            statsResponse.player.achievements.skyblock_dungeoneer
+                                    > NoScamSpam.config.catacombsLevel) {
+                        Minecraft.getMinecraft().thePlayer.addChatMessage(event.message);
+                        NoScamSpam.cachedAllowed.add(username);
+                        ClientChatReceivedEvent newEvent = new ClientChatReceivedEvent((byte) 0, event.message);
+                        MinecraftForge.EVENT_BUS.post(newEvent);
```

Comment:
Is there any benefit to canceling the event then reposting it rather than just not canceling it at all? This ties into adding the chat message via `addChatMessage` versus just letting it fall through (see the other review comment below).

Reply:
Yes becuse the forge events stop the game loop until processing is completed.